AutoSitemap_body.php

```diff
@@ -89,34 +89,88 @@
 
 if (!isset($wgAutoSitemap["footer"]           )) $wgAutoSitemap["footer"]              = "\n</urlset>";
 
+if (!isset($wgAutoSitemap["min_age"]          )) $wgAutoSitemap["min_age"]             = 3600; // One hour in seconds
+
 class AutoSitemap {
     static public function writeSitemap() {
         global $wgAutoSitemap;
 
+        $filename = $wgAutoSitemap["filename"];
+        $min_age = $wgAutoSitemap["min_age"];
+
+        $mtime = filemtime($filename);
+        if ($mtime !== FALSE && time() - $mtime < $min_age) {
+           // Sitemap is young, no need to update.
+           return;
+        }
+
         $server       = $wgAutoSitemap["server"];
-        $filename     = $wgAutoSitemap["filename"];
         $tmp_filename = $filename.'.tmp'.bin2hex(random_bytes(16)).'.tmp';
 
-        $file_handle = fopen($tmp_filename, 'w') or die('Cannot write to '.$tmp_filename.'.');
+        $file_handle = fopen($tmp_filename, 'w');
+        if ($file_handle === FALSE) {
+           error_log("Couldn't fopen file: $tmp_filename");
+           return;
+        }
 
         $dbr = wfGetDB(DB_REPLICA);
         $res = $dbr->query(self::getSQL());
 
-        $count = $res->numRows();
-        $pos   = 0;
-
-        $data = $wgAutoSitemap["header"];
-        while($row = $res->fetchObject()) {
-            $data .= self::formatResult($server, $row, $pos, $count);
-            ++$pos;
+        // Sitemaps are limited to 50,000 URLs and 50 MB.
+        $entries    = 0;
+        $bytes      = 0;
+        $maxEntries = 50000;
+        $maxBytes   = 50000000;
+
+        /* Note:
+         *
+         * Actually, https://www.sitemaps.org/protocol.html says the limit is 52,428,800 bytes,
+         * which is 50 MiB (MiB = 1024 * 1024), but it uses the old notation of "MB" which has
+         * since been standardized in ISO to mean 1000 * 1000.  We use the lower limit just in
+         * case, because some systems may interpret the "50MB" to mean 50 * 1000 * 1000.
+         *
+         * It's very unlikely to reach the size limit before reaching the 50k URL limit anyway,
+         * unless there are tons of *extremely* long URLs in a sitemap.
+         */
+
+        $error = FALSE;
+
+        try {
+            $bytes += self::write($file_handle, $wgAutoSitemap["header"]);
+            while($row = $res->fetchObject()) {
+                if ($entries == $maxEntries) {
+                    throw new Exception("Sitemap URL limit exceeded, aborting.");
```

Comment:
Exceeding the limit leads to creating file without footer

Reply:
The tmp file will be deleted since `$error` is set to true in the `catch` block.